internal/backend/sftp/sftp.go

```diff
@@ -258,23 +258,40 @@ func (r *SFTP) Save(ctx context.Context, h restic.Handle, rd restic.RewindReader
 	}
 
 	filename := r.Filename(h)
+	dirname := r.Dirname(h)
 
-	// create new file
-	f, err := r.c.OpenFile(filename, os.O_CREATE|os.O_EXCL|os.O_WRONLY)
+	// Ensure that the containing directory exists.
+	if err := r.c.MkdirAll(dirname); err != nil {
+		debug.Log("error creating dir %v: %v", r.Dirname(h), err)
+		return errors.Wrap(err, "Mkdir")
+	}
 
-	if r.IsNotExist(err) {
-		// error is caused by a missing directory, try to create it
-		mkdirErr := r.c.MkdirAll(r.Dirname(h))
-		if mkdirErr != nil {
-			debug.Log("error creating dir %v: %v", r.Dirname(h), mkdirErr)
-		} else {
-			// try again
-			f, err = r.c.OpenFile(filename, os.O_CREATE|os.O_EXCL|os.O_WRONLY)
-		}
+	// Try and see if there is space left for this file.
+	// We do this because the OpenSSH sftp-server only returns
+	// a generic "failure" code when the filesystem is full,
+	// so detecting after the fact is impossible.
+	if err := r.checkSpace(dirname, rd.Length()); err != nil {
+		return err
 	}
 
+	f, err := r.c.OpenFile(filename, os.O_CREATE|os.O_EXCL|os.O_WRONLY)
 	if err != nil {
-		return errors.Wrap(err, "OpenFile")
+		return err
+	}
+	defer func() {
+		// Try not to leave a partial file behind.
+		if err == nil {
+			return
+		}
+		err := r.c.Remove(f.Name())
+		if err != nil {
+			debug.Log("sftp failed to remove broken file %v: %v", filename, err)
+		}
+	}()
+
+	// pkg/sftp doesn't allow creating with a mode.
+	if err = f.Chmod(backend.Modes.File); err != nil {
```

Comment:
What is the reason for changing the file permissions earlier on (just curious)?

Reply:
If we set them later, remote users may see what is written to the backup file while we're busy (depending on the remote umask).